scripts/mayachemtools/docs/modules/man3/Fsp3CarbonsDescriptors.3
.\" Automatically generated by Pod::Man 2.28 (Pod::Simple 3.35)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "Fsp3CarbonsDescriptors 1"
.TH Fsp3CarbonsDescriptors 1 "2018-05-15" "perl v5.22.4" "MayaChemTools"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
Fsp3CarbonsDescriptors
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
use MolecularDescriptors::Fsp3CarbonsDescriptors;
.PP
use MolecularDescriptors::Fsp3CarbonsDescriptors qw(:all);
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBFsp3CarbonsDescriptors\fR class provides the following methods:
.PP
new, GenerateDescriptors, GetDescriptorNames, StringifyFsp3CarbonsDescriptors
.PP
\&\fBFsp3CarbonsDescriptors\fR is derived from \fBMolecularDescriptors\fR class which in turn
is  derived from \fBObjectProperty\fR base class that provides methods not explicitly defined
in \fBFsp3CarbonsDescriptors\fR, \fBMolecularDescriptors\fR or \fBObjectProperty\fR classes using Perl's
\&\s-1AUTOLOAD\s0 functionality. These methods are generated on-the-fly for a specified object property:
.PP
.Vb 3
\&    Set<PropertyName>(<PropertyValue>);
\&    $PropertyValue = Get<PropertyName>();
\&    Delete<PropertyName>();
.Ve
.PP
Fraction sp3 carbons (Fsp3Carbons) [ Ref 115\-116, Ref 119 ] value is calculated by dividing the number of sp3
carbons (Sp3Carbons) with the total number of carbons in a molecule.
.SS "\s-1METHODS\s0"
.IX Subsection "METHODS"
.IP "\fBnew\fR" 4
.IX Item "new"
.Vb 2
\&    $NewFsp3CarbonsDescriptors = new MolecularDescriptors::
\&                                 Fsp3CarbonsDescriptors(%NamesAndValues);
.Ve
.Sp
Using specified \fIFsp3CarbonsDescriptors\fR property names and values hash, \fBnew\fR
method creates a new object and returns a reference to newly created \fBFsp3CarbonsDescriptors\fR
object. By default, the following properties are initialized:
.Sp
.Vb 4
\&    Molecule = \*(Aq\*(Aq
\&    Type = \*(AqFsp3Carbons\*(Aq
\&    @DescriptorNames = (\*(AqFsp3Carbons\*(Aq, \*(AqSp3Carbons\*(Aq)
\&    @DescriptorValues = (\*(AqNone\*(Aq, \*(AqNone\*(Aq)
.Ve
.Sp
Examples:
.Sp
.Vb 2
\&    $Fsp3CarbonsDescriptors = new MolecularDescriptors::Fsp3CarbonsDescriptors(
\&                              \*(AqMolecule\*(Aq => $Molecule);
\&
\&    $Fsp3CarbonsDescriptors = new MolecularDescriptors::Fsp3CarbonsDescriptors();
\&
\&    $Fsp3CarbonsDescriptors\->SetMolecule($Molecule);
\&    $Fsp3CarbonsDescriptors\->GenerateDescriptors();
\&    print "Fsp3CarbonsDescriptors: $Fsp3CarbonsDescriptors\en";
.Ve
.IP "\fBGenerateDescriptors\fR" 4
.IX Item "GenerateDescriptors"
.Vb 1
\&    $Fsp3CarbonsDescriptors\->GenerateDescriptors();
.Ve
.Sp
Calculates Fsp3Carbons and Sp3Carbons values for a molecule and returns \fIFsp3CarbonsDescriptors\fR.
.IP "\fBGetDescriptorNames\fR" 4
.IX Item "GetDescriptorNames"
.Vb 3
\&    @DescriptorNames = $Fsp3CarbonsDescriptors\->GetDescriptorNames();
\&    @DescriptorNames = MolecularDescriptors::Fsp3CarbonsDescriptors::
\&                       GetDescriptorNames();
.Ve
.Sp
Returns all available descriptor names as an array.
.IP "\fBStringifyFsp3CarbonsDescriptors\fR" 4
.IX Item "StringifyFsp3CarbonsDescriptors"
.Vb 1
\&    $String = $Fsp3CarbonsDescriptors\->StringifyFsp3CarbonsDescriptors();
.Ve
.Sp
Returns a string containing information about \fIFsp3CarbonsDescriptors\fR object.
.SH "AUTHOR"
.IX Header "AUTHOR"
Manish Sud <msud@san.rr.com>
.SH "SEE ALSO"
.IX Header "SEE ALSO"
MolecularDescriptors.pm, MolecularDescriptorsGenerator.pm
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright (C) 2018 Manish Sud. All rights reserved.
.PP
This file is part of MayaChemTools.
.PP
MayaChemTools is free software; you can redistribute it and/or modify it under
the terms of the \s-1GNU\s0 Lesser General Public License as published by the Free
Software Foundation; either version 3 of the License, or (at your option)
any later version.


scripts/mayachemtools/lib/MolecularDescriptors/Fsp3CarbonsDescriptors.pm
package MolecularDescriptors::Fsp3CarbonsDescriptors;
#
# File: Fsp3CarbonsDescriptors.pm
# Author: Manish Sud <msud@san.rr.com>
#
# Copyright (C) 2018 Manish Sud. All rights reserved.
#
# This file is part of MayaChemTools.
#
# MayaChemTools is free software; you can redistribute it and/or modify it under
# the terms of the GNU Lesser General Public License as published by the Free
# Software Foundation; either version 3 of the License, or (at your option) any
# later version.
#
# MayaChemTools is distributed in the hope that it will be useful, but without
# any warranty; without even the implied warranty of merchantability of fitness
# for a particular purpose.  See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with MayaChemTools; if not, see <http://www.gnu.org/licenses/> or
# write to the Free Software Foundation Inc., 59 Temple Place, Suite 330,
# Boston, MA, 02111-1307, USA.
#

use strict;
use Carp;
use Exporter;
use Scalar::Util ();
use TextUtil ();
use MathUtil ();
use Atom;
use Molecule;
use MolecularDescriptors::MolecularDescriptors;

use vars qw(@ISA @EXPORT @EXPORT_OK %EXPORT_TAGS);

@ISA = qw(MolecularDescriptors::MolecularDescriptors Exporter);
@EXPORT = qw();
@EXPORT_OK = qw(GetDescriptorNames);

%EXPORT_TAGS = (all  => [@EXPORT, @EXPORT_OK]);

# Setup class variables...
my($ClassName, @DescriptorNames);
_InitializeClass();

# Overload Perl functions...
use overload '""' => 'StringifyFsp3CarbonsDescriptors';

# Class constructor...
sub new {
  my($Class, %NamesAndValues) = @_;

  # Initialize object...
  my $This = $Class->SUPER::new();
  bless $This, ref($Class) || $Class;
  $This->_InitializeFsp3CarbonsDescriptors();

  $This->_InitializeFsp3CarbonsDescriptorsProperties(%NamesAndValues);

  return $This;
}

# Initialize class ...
sub _InitializeClass {
  #Class name...
  $ClassName = __PACKAGE__;

  # Descriptor names...
  @DescriptorNames = ('Fsp3Carbons', 'Sp3Carbons');

}

# Get descriptor names as an array.
#
# This functionality can be either invoked as a class function or an
# object method.
#
sub GetDescriptorNames {
  return @DescriptorNames;
}

# Initialize object data...
#
sub _InitializeFsp3CarbonsDescriptors {
  my($This) = @_;

  # Type of MolecularDescriptor...
  $This->{Type} = 'Fsp3Carbons';

  # Intialize descriptor names and values...
  $This->_InitializeDescriptorNamesAndValues(@DescriptorNames);

  return $This;
}

# Initialize object properties...
#
sub _InitializeFsp3CarbonsDescriptorsProperties {
  my($This, %NamesAndValues) = @_;

  my($Name, $Value, $MethodName);
  while (($Name, $Value) = each  %NamesAndValues) {
    $MethodName = "Set${Name}";
    $This->$MethodName($Value);
  }

  return $This;
}

# Calculate fraction of SP3 carbons (Fsp3Carbons)  [ Ref 115-116, Ref 119 ] in a molecule...
#
# It is defined as follows:
#
# Fsp3 = Number of SP3 carbons/Total number of carbons
#
sub GenerateDescriptors {
  my($This) = @_;

  # Initialize descriptor values...
  $This->_InitializeDescriptorValues();

  # Check availability of molecule...
  if (!$This->{Molecule}) {
    carp "Warning: ${ClassName}->GenerateDescriptors: $This->{Type} molecular descriptors generation didn't succeed: Molecule data is not available: Molecule object hasn't been set...";
    return undef;
  }

  # Calculate descriptor values...
  if (!$This->_CalculateDescriptorValues()) {
    carp "Warning: ${ClassName}->GenerateDescriptors: $This->{Type} molecular descriptors generation didn't succeed: Couldn't calculate Fsp3Carbons values corresponding to assigned Fsp3Carbons atom types...";
    return undef;
  }

  # Set final descriptor values...
  $This->_SetFinalDescriptorValues();

  return $This;
}

# Calculate Fsp3Carbons value...
#
sub _CalculateDescriptorValues {
  my($This) = @_;
  my($Atom, $AtomID, $TotalCarbons, $NumOfSp3Carbons, $Fsp3Carbons);

  $TotalCarbons = 0;
  $NumOfSp3Carbons = 0;

  ATOM: for $Atom ($This->{Molecule}->GetAtoms()) {
    if (!$Atom->IsCarbon()) {
      next ATOM;
    }
    $TotalCarbons += 1;

    if ($Atom->DoesAtomNeighborhoodMatch('C.T4.TSB4')) {
      $NumOfSp3Carbons += 1;
    }
  }

  $Fsp3Carbons = $NumOfSp3Carbons ? $NumOfSp3Carbons/$TotalCarbons : 0;

  # Track values...
  $This->{Fsp3Carbons} = MathUtil::round($Fsp3Carbons, 2);
  $This->{Sp3Carbons} = $NumOfSp3Carbons;

  return $This;
}

# Setup final descriptor values...
#
sub _SetFinalDescriptorValues {
  my($This) = @_;

  $This->{DescriptorsGenerated} = 1;

  $This->SetDescriptorValues($This->{Fsp3Carbons}, $This->{Sp3Carbons});

  return $This;
}

# Return a string containg data for Fsp3CarbonsDescriptors object...
#
sub StringifyFsp3CarbonsDescriptors {
  my($This) = @_;
  my($Fsp3CarbonsDescriptorsString);

  $Fsp3CarbonsDescriptorsString = "MolecularDescriptorType: $This->{Type}; " . $This->_StringifyDescriptorNamesAndValues();

  return $Fsp3CarbonsDescriptorsString;
}

# Is it a Fsp3CarbonsDescriptors object?
sub _IsFsp3CarbonsDescriptors {
  my($Object) = @_;

  return (Scalar::Util::blessed($Object) && $Object->isa($ClassName)) ? 1 : 0;
}

1;

__END__

=head1 NAME

Fsp3CarbonsDescriptors

=head1 SYNOPSIS

use MolecularDescriptors::Fsp3CarbonsDescriptors;

use MolecularDescriptors::Fsp3CarbonsDescriptors qw(:all);

=head1 DESCRIPTION

B<Fsp3CarbonsDescriptors> class provides the following methods:

new, GenerateDescriptors, GetDescriptorNames, StringifyFsp3CarbonsDescriptors

B<Fsp3CarbonsDescriptors> is derived from B<MolecularDescriptors> class which in turn
is  derived from B<ObjectProperty> base class that provides methods not explicitly defined
in B<Fsp3CarbonsDescriptors>, B<MolecularDescriptors> or B<ObjectProperty> classes using Perl's
AUTOLOAD functionality. These methods are generated on-the-fly for a specified object property:

    Set<PropertyName>(<PropertyValue>);
    $PropertyValue = Get<PropertyName>();
    Delete<PropertyName>();

Fraction sp3 carbons (Fsp3Carbons) [ Ref 115-116, Ref 119 ] value is calculated by dividing the number of sp3
carbons (Sp3Carbons) with the total number of carbons in a molecule.

=head2 METHODS

=over 4

=item B<new>

    $NewFsp3CarbonsDescriptors = new MolecularDescriptors::
                                 Fsp3CarbonsDescriptors(%NamesAndValues);

Using specified I<Fsp3CarbonsDescriptors> property names and values hash, B<new>
method creates a new object and returns a reference to newly created B<Fsp3CarbonsDescriptors>
object. By default, the following properties are initialized:

    Molecule = ''
    Type = 'Fsp3Carbons'
    @DescriptorNames = ('Fsp3Carbons', 'Sp3Carbons')
    @DescriptorValues = ('None', 'None')

Examples:

    $Fsp3CarbonsDescriptors = new MolecularDescriptors::Fsp3CarbonsDescriptors(
                              'Molecule' => $Molecule);

    $Fsp3CarbonsDescriptors = new MolecularDescriptors::Fsp3CarbonsDescriptors();

    $Fsp3CarbonsDescriptors->SetMolecule($Molecule);
    $Fsp3CarbonsDescriptors->GenerateDescriptors();
    print "Fsp3CarbonsDescriptors: $Fsp3CarbonsDescriptors\n";

=item B<GenerateDescriptors>

    $Fsp3CarbonsDescriptors->GenerateDescriptors();

Calculates Fsp3Carbons and Sp3Carbons values for a molecule and returns I<Fsp3CarbonsDescriptors>.

=item B<GetDescriptorNames>

    @DescriptorNames = $Fsp3CarbonsDescriptors->GetDescriptorNames();
    @DescriptorNames = MolecularDescriptors::Fsp3CarbonsDescriptors::
                       GetDescriptorNames();

Returns all available descriptor names as an array.

=item B<StringifyFsp3CarbonsDescriptors>

    $String = $Fsp3CarbonsDescriptors->StringifyFsp3CarbonsDescriptors();

Returns a string containing information about I<Fsp3CarbonsDescriptors> object.

=back

=head1 AUTHOR

Manish Sud <msud@san.rr.com>

=head1 SEE ALSO

MolecularDescriptors.pm, MolecularDescriptorsGenerator.pm

=head1 COPYRIGHT

Copyright (C) 2018 Manish Sud. All rights reserved.

This file is part of MayaChemTools.

MayaChemTools is free software; you can redistribute it and/or modify it under
the terms of the GNU Lesser General Public License as published by the Free
Software Foundation; either version 3 of the License, or (at your option)
any later version.

=cut
